ibm/service/power/resource_ibm_pi_network_address_group.go

```diff
@@ -0,0 +1,247 @@
+				Description: "The user tags associated with this resource.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Optional:    true,
+				Set:         schema.HashString,
+				Type:        schema.TypeSet,
+			},
+			// Attributes
+			Attr_CRN: {
+				Computed:    true,
+				Description: "The Network Address Group's crn.",
+				Type:        schema.TypeString,
+			},
+			Attr_Members: {
+				Computed:    true,
+				Description: "The list of IP addresses in CIDR notation (for example 192.168.66.2/32) in the Network Address Group.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						Attr_CIDR: {
+							Computed:    true,
+							Description: "The IP addresses in CIDR notation for example 192.168.1.5/32.",
+							Type:        schema.TypeString,
+						},
+						Attr_ID: {
+							Computed:    true,
+							Description: "The id of the Network Address Group member IP addresses.",
+							Type:        schema.TypeString,
+						},
+					},
+				},
+				Optional: true,
+				Type:     schema.TypeList,
+			},
+			Attr_NetworkAddressGroupID: {
+				Computed:    true,
+				Description: "The unique identifier of the network address group.",
+				Type:        schema.TypeString,
+			},
+		},
+	}
+}
+
+func resourceIBMPINetworkAddressGroupCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := d.Get(Arg_CloudInstanceID).(string)
+	name := d.Get(Arg_Name).(string)
+	nagC := instance.NewIBMPINetworkAddressGroupClient(ctx, sess, cloudInstanceID)
+	var body = &models.NetworkAddressGroupCreate{
+		Name: &name,
+	}
+
+	if v, ok := d.GetOk(Arg_UserTags); ok {
+		body.UserTags = flex.ExpandStringList(v.([]interface{}))
+	}
+
+	networkAddressGroup, err := nagC.Create(body)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if _, ok := d.GetOk(Arg_UserTags); ok {
+		if networkAddressGroup.Crn != nil {
+			oldList, newList := d.GetChange(Arg_UserTags)
+			err := flex.UpdateGlobalTagsUsingCRN(oldList, newList, meta, string(*networkAddressGroup.Crn), "", UserTagType)
+			if err != nil {
+				log.Printf("Error on update of pi network address group (%s) pi_user_tags during creation: %s", *networkAddressGroup.ID, err)
+			}
+		}
+	}
+	d.SetId(fmt.Sprintf("%s/%s", cloudInstanceID, *networkAddressGroup.ID))
+
+	return resourceIBMPINetworkAddressGroupRead(ctx, d, meta)
+}
+
+func resourceIBMPINetworkAddressGroupRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID, nagID, err := splitID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	nagC := instance.NewIBMPINetworkAddressGroupClient(ctx, sess, cloudInstanceID)
+	networkAddressGroup, err := nagC.Get(nagID)
+	if err != nil {
+		return diag.FromErr(err)
```

Comment:
We need to handle scenario when user delete a resource outside Terraform context .
Earlier we were handling this behavior by adding Exists method but now we can handle as part of Read itself .
If get returns 404 then d.SetID to empty.

I think we need to handle this across all power resources because its missing the Exists (old behavior) or (new behaviour)
sample code
```
vpnServerRoute, response, err := sess.GetVPNServerRouteWithContext(context, getVPNServerRouteOptions)
	if err != nil {
		if response != nil && response.StatusCode == 404 {
			d.SetId("")
			return nil
		}
		log.Printf("[DEBUG] GetVPNServerRouteWithContext failed %s\n%s", err, response)
		return diag.FromErr(fmt.Errorf("[ERROR] GetVPNServerRouteWithContext failed %s\n%s", err, response))
	}
	```

Reply:
Agreed, but when I was working with PowerVS, 404 was not implemented for all APIs, and I am sure it is still the same.
Secondly, the client calls don't return the response status code, we wrap the error messages. In Terraform we unwrap the error and find 404 in it for a few APIs already.

Reply:
Updated